mantidproject/mantid
# Fix copy/paste bugs in reflectometry GUI

**Description of work.**

This PR fixes the following bugs to do with cut/copy/paste in the reflectometry GUI:
- Cutting and pasting rows can cause crashing because the GUI is updated but the model is not.
- Pasting onto an incorrect depth can cause crashing because an unexpected number of cells exists on the row.
- Pasting a group should create a unique group name or we can get unexpected behaviour because of assumptions that non-empty group names are unique.

A new `Clipboard` class has been added to track not only the selected subtrees (which contain relative paths to rows/groups on the clipboard relative to one or more "roots") but also the roots themselves so that we can tell whether a row or group was copied. This is so that we can disable pasting a row onto a group and vice versa. We can also make assumptions about the content of the subtrees when we know what level we're dealing with:
- If any of the selected roots is a group, then all roots must be groups and each one can contain zero or more rows.
- If none of the selected roots is a group then all of the roots must be rows.

Note that unit tests have not been added yet because this is a bigger job because there are currently no tests at all for cutting and pasting. This is noted on #23027 and I'll do this as a separate PR.

**To test:**

- Open the `ISIS Reflectometry` GUI
- Try as many combinations of cut/copy/paste of rows and/or groups as you can think of.
- Ensure the GUI does something sensible and does not crash.

Refs #23027 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
